### ISSUE TITLE
show routed messages to user

### ISSUE THREAD

**Issue by huss** [OpenEnergyDashboard/OED#869]
### Is your feature request related to a problem? Please describe.

PR #857 sends back messages from the conversion and unit routes in the desired way (code in src/server/routes/conversions.js or units.js). The API returns what is routed (e.g., src/client/app/utils/api/ConversionsApi.ts). The action that dispatches the code ignores the returned values (e.g., src/client/app/actions/conversions.ts) and just puts up generic messages. It would be nice to display the returned messages. Note it is often only done for failures as success does not return a value.

Examples of code already doing this are:

- src/client/app/actions/meters.ts in submitEditedMeter() for failure.
- src/client/app/components/csv/ReadingsCSVUploadComponent.tsx in handleSubmit() displays the message on success.

### Describe the solution you'd like

The message returned from the route should be passed up the client call chain to an appropriate point to display. Note some code uses a window alert but the showSuccessNotification/showErrorNotification now have a time argument to control how long it stays up so that would probably be better. For some error messages where the user needs lots of time to read or wants to copy, then getting it to stay up until closed would be nice. I also have not checked how easy it is to copy the displayed text on these messages.

### Additional context

There are other places with notices that could be investigated for similar changes.

**Comment by ArtP1** [comment]
I'm working on this

**Comment by huss** [comment]
Given OED now allows admins to easily see log messages, the full messages are now available in the log. OED should limit the content of messages to explain what happened and not, in general, show the internal details of the issue. The message could tell users to see the log or contact the site admin for help.

**Comment by aduques** [comment]
Referencing work done in Issue #1191.

Since I have added a disabled feature that disables the save button if changes are not valid, I'm not sure how the failure messages can be reachable. However, I did notice that the success messages were not implemented everywhere.

Note: While the known ways of the failure messages have been essentially prevented by the implementation of the unsaved warning, failure/success messages should be used everywhere when applicable, so that features can be debugged in the future if they ever fail.

Files that don't have a success/failure message present:
CreateConversionComponent.tsx
EditConversionComponent.tsx
CreateGroupModalComponent.tsx
EditGroupModalComponent.tsx

Could not check completely:
MetersCSVUploadComponent.tsx
ReadingsCSVUploadComponent.tsx

**Comment by aduques** [comment]
I apologize for the delay on starting this issue. Before I begin, I want to clarify the desired behavior of this issue. As previously discussed, this issue lined up with the previous issues that I've worked on, where I worked with the edit/create pages. The goal of this issue is to:

- For all edit/create pages, implement a success/failure notification that displays specific information of the selected edit/create. This is implemented on some of the edit/create pages, so I will make sure that all will have it.
- Failure notifications will be hard to encounter naturally due to the handleSave/canSave (feature that disables saves for invalid edits), but the success/failure notification is still desirable to have for future work.

If the above is the only desired, I think that the approach to the solution will be relatively easy as it would just involve working on the edit/create pages and finding the file where the success/failure notifications are initialized and customized. 

However, I am anticipating one possible problem to the solution. In order to test the visibility of the failure notifications, the only option that comes to mind is to disable the handleSave feature on the edit/create pages as I work and re-enable it at the end. This approach would be prone to bugs if not handled correctly (ex. forgetting to re-enable on some files), so I would appreciate any ideas to test the changes.

**Comment by huss** [comment]
> However, I am anticipating one possible problem to the solution. In order to test the visibility of the failure notifications, the only option that comes to mind is to disable the handleSave feature on the edit/create pages as I work and re-enable it at the end. This approach would be prone to bugs if not handled correctly (ex. forgetting to re-enable on some files), so I would appreciate any ideas to test the changes.

You are correct that extra steps are needed to verify that failure works correctly. I agree that no matter how you do this it is going to require adding special code and then removing it. How you track the code is up to you. One technique I use is to put a comment on the code with ``// TODO DEBUG ......`` since a comment starting with TODO can automatically be highlighted if you have the proper extension (search for todo in the VSC extensions to find a good one). I use two techniques to make sure I remove them all when done:

1. Search for TODO DEBUG to find any left.
2. I check all files for all changes before I commit them to make sure nothing is left over. You can also check the changes that will go into the pull request before it is made. Let me know if you want advice on how to do this. Personally I try to do both for any commit or PR to be careful.

Removing the disable on save is certainly an option to get failures. Here are a couple of other thoughts:

- Modify the data to be sent to the server route on the client. This can be done by modifying the data in the component right before it is sent to the route in the save or in the actual client-side route. I usually pick something easy such as a bad argument value (causes the validation to fail) and/or a duplicate key (name, id, ...) to cause a foreign key failure in the database. Doing both shows both types of errors will be shown.
- Modify the route on the server to modify the values before they are actually used. This will impact all client uses but that probably is less important to you since I think each route is used on one page. This can be done before validation to make that error happen and after for a DB error. This is in a separate file so requires touching it but can be easier to undo if no other changes were made to the file since you can simply revert in git.

There are other ways but these are ones I have certainly used in the past. Please let me know if anything is not clear or you need anything else.

**Comment by aduques** [comment]
FOUND BUGS:
EditUnitModalComponent.tsx (possibly in other files)

- Once the canSave check is removed and invalid changes are attempted, NO error notifications is displayed despite being implemented (showErrorNotifications was initially already implemented here) AND after a failed save the existing saved values of the selected unit is not saved. When saving on the UnsavedWarning’s ‘Save All’ with an invalid change pending, the save all goes through but reverts the changes back to the initial state.

- (If you clear the ‘Name’ field → try to save → save failed → no notification → try and edit the same meter → the ‘Name’ field is still blank → try to click out → Unsaved Warning appears → press ‘Save All’ on Unsaved Warning with ‘Name’ field blank still → save goes through, but reverts back to the initially saved values)

QUESTION: In addition to fixing the visibility of the Error Notification, should the above functionality be looked at also? The above behavior would not be repeatable once the canSave check is re-enabled. A potential fix could be to make sure that resetState is called even after a failed save.

Update on progress: I have currently only looked at implementing the Success and Error notifications on the Unit modals. The above behavior has halted me from testing the notifications on the other create/edit pages as I want to be able to successfully get one page to work as expected before moving on. 

In addition, to update my initial understanding of the issue, despite some of the create/edit pages having an implementation for ShowSuccessNotification and ShowErrorNotification, the notifications are not being displayed. Therefore, I need to look into as why this is happening.

Also, I am acknowledging here that in addition to reimplementing the notifications on edit/create pages, it is also desired in the issue to implement a timer feature for these notifications. The notifications would disappear automatically after a set time, instead of only being manually closed. However with the current issues that I’m facing, this feature will be put off till after notifications can be implemented.

Summary: unexpected behavior with saving with invalid changes, error notifications not showing despite being implemented, acknowledging timer feature.

**Comment by huss** [comment]
@aduques Thank you for your message and careful checking. I was curious/concerned about not getting failure notices so I tried to reproduce in a different way. Here is what I did:

1. Inside of the handleSaveChanges in src/client/app/components/unit/EditUnitModalComponent.tsx right before it is submitted I did ``submitState.secInRate = -1;``. This makes the value illegal by a database check.

<img width="839" height="857" alt="Image" src="https://github.com/user-attachments/assets/9f05959f-8e97-4d49-8731-86dba87983d3" />

2. I used the web browser on the unit edit page to change unit (so save was enabled) and saved. In my case I changed the seconds in rate to another valid value but that is not be important. I then saw this notice:

<img width="1295" height="942" alt="Image" src="https://github.com/user-attachments/assets/e803a8ad-f6cf-4b49-9438-b15b4ac9dd9f" />

and the terminal where I started OED shows this message as expected:

<img width="1765" height="340" alt="Image" src="https://github.com/user-attachments/assets/4c158d29-7dc1-4207-97d9-7cf3ec0dff67" />

The value was not edited and shows the old value on the edit unit page (before and after refreshing the web browser page). To be careful, I changed my new line of code to ``submitState.secInRate = 1;``. After that the save works and the value is set to 1 no matter what I set it to before the save. I also get a success notice. Thus, the error seems to show and work for me. Could you try this to see if it works for you? If so, I'm curious why your change did not work. If you don't know then maybe provide me the code to check that out. If not, we need to look into this more to figure out why we see different things.

Now for your questions about auto dismissing the notification of error. src/client/app/utils/notifications.ts has the functions used to display messages. For errors it has ``export function showErrorNotification(message: string, position: ToastPosition = toast.POSITION.TOP_RIGHT, autoDismiss = 15000) {`` so it will close (dismiss) the popup message after 15000 ms or 15 seconds. In general, OED code should not set this value unless there is a special reason for several reasons:

1. It brings consistency to the UX.
2. It allows for encapsulation so OED can easily change the behavior for all uses.

Currently error messages stay up longer than other notices. The logic is they are more important. The user can still click the "x" on the message to close earlier. If you think the autoDismiss value should be different then please let us know so we can consider that change.

Overall, the dismiss should be happening automatically so I think this aspect is already complete. I do note that if you leave the web browser while the message is showing and go to another app then the auto dismiss may not happen. I believe this is a common feature that once the focus is gone on that web page it is not doing actions. I don't think this is a major issue and is fine. The issue was noting that the proposed system for messages had this feature.

Please let me know if you have any thoughts or questions.

**Comment by aduques** [comment]
Thank you @huss for the detailed steps of how to reproduce the error notifications. I initially attempted to bypass the prevention of saving invalid changes by doing the following:

			{/* Unsaved Warning Component */}
			{showUnsavedWarning && (
				<SimpleUnsavedWarningComponent
					isOpen={showUnsavedWarning}
					onDiscard={() => {
						setShowUnsavedWarning(false);
						setHasUnsavedChanges(false);
						handleClose();
						resetState();
					}}
					onConfirm={() => {
						setShowUnsavedWarning(false);
						setHasUnsavedChanges(false);
						handleSaveChanges();
						handleClose();
					}}
					onCancel={() => setShowUnsavedWarning(false)}
					// TODO DEBUG: Disabled is always set to false,
					// which ignores the cases of invalid saves being rejected
					// This is needed in order to test the showErrorNotification
					//disabled={!canSave}
					disabled={false}
				/>
			)}

All related code to canSave were commented out in the same manner.

With the above approach, I was able to attempt save, but there was no indication of an error notification. I would assume that the current implementation of the error notification does not take into account of invalid changes (canSave boolean), meaning that the error notification did not appear as it did not detect a case where it needed to appear (the error notification implementation has not been updated to work with the unsaved warnings). I will spend time to investigate the functionality of the notification. 

I will most likely revert this approach and try out the ones that you have recommended to see if I can reproduce the error notification.

I appreciate the information pointing to an existing implementation of the timer information. I had not known that the error notifications would already disappear after a set time as I usually closed out of them relatively fast. I will keep these new information in mind as I continue working.

**Comment by huss** [comment]
@aduques Thanks for the clear explanation of what you did. I retried with my code where it sets the sec. in rate to a bad value. I changed the note so it would enable save. I clicked off the modal and received the unsaved warning modal. I then clicked save on that and it generated the error message as with the save button the edit unit page. This is what I would have expected because, as your image above shows, onConfirm calls handleSaveChanges() which properly handles errors. This is a good design as it reused the original method.

Given this, I'm unclear on why you don't see the error. Is there any chance that you did not submit an invalid unit so it was not rejected?

**Comment by aduques** [comment]
@huss I have only tried saving an edit when the 'Name' has been set to be emptied as it is a known required attribute for the Units. I was assuming that if this field were to be left empty when attempting a save, it would result in an error notification. I am in the process of following your steps to reproduce the error.

I was assuming that my initial plan of testing the showErrorNotification was incorrect as I was never testing invalid units.

EDIT: I have added the 'submitState.secInRate = -1;' line and was able to reproduce the error notification. So I am assuming that I was never initially testing invalid units. Though, I am still unclear as to why forcing the save all button to be enabled at all times, despite the validity of the unit, does not produce the error notification.

EDIT: I was able to replicate the same behavior for CreateUnitModalComponent.tsx by using the same test. I appreciate the help of finding a valid way to test the error/success notifications, and will use a similar approach for the other create/edit pages now.

**Comment by huss** [comment]
@aduques I agree that an empty name should cause and error since the units table has:
```name VARCHAR(50) UNIQUE NOT NULL CHECK (char_length(name) >= 1)```
I do see if the name is set but not the identifier then it uses the name as the identifier. Any chance you set the name but not the identifier? If not, I'll try to see why it would not cause an error.

**Comment by aduques** [comment]
@huss I assume that only setting the 'Name' field to blank while not modifying the 'Identifier' field is the reason why the blank 'Name' field change is not flagging an error.

		// Only validate and store if any changes.
		if (meterHasChanges) {
			// Set default identifier as name if left blank
			localMeterEdits.identifier = (!localMeterEdits.identifier || localMeterEdits.identifier.length === 0) ?
				localMeterEdits.name : localMeterEdits.identifier;

Snippet from handleSaveChanges() in EditMeterModalComponent.tsx

Therefore, to test the showErrorNotification, I included the following lines:
				// TODO DEBUG: added in to test the showErrorNotification
				submitState.name = '';
				submitState.identifier = '';

Clearing only the identifier results in the name autofilling for it, so both fields need to be cleared in order to view the error notification.

**Comment by huss** [comment]
I looked into this. 

- Setting ``submitState.name = ''`` right before submitEditedUnit does cause an error and the terminal indicates that it was that name is not at least one character. I then remove this so as not to modify the name.
- Changing the button code to ``<Button color='primary' onClick={handleSaveChanges}>`` so not to disable does not cause an error on empty name edit. This is the reason:
  - handleSaveChanges calls shouldUpdateUnit. This finds inputOk to be true so it returns canSave which is false. As a result, the save does not happen. This means that enabling the save button does not actually allow the save.
  - I modified shouldUpdateUnit to return true instead of canSave. It then generates an error if you remove the name characters during save.

Given this, I think the code is doing what it should and the error is working as intended. shouldUpdateUnit is doing additional checks but it must return false if the save is not allowed. Now, the user could not click save if canSave was false so that is technically not essential that it return canSave but costs nothing and is a good sanity check. I suspect that this fact is why you did not see the error and the logic was unexpected. Please let me know if you have any thoughts or see something different.

**Comment by aduques** [comment]
Here is an update on the status of the create/edit pages:

Implemented, but not specific messages (generic messages)

- CreateUnitModalComponent.tsx
  - Message: "Failed to create unit" 
- EditUnitModalComponent.tsx
  - Message: "Failed to edit unit" 
- PreferencesComponent.tsx
  - NOTE: Message: “Updated preferences”; Also, not sure how to test the failure message here as the implementation on Admin Settings used the old UnsavedWarningComponent which was hardcoded to work with PreferencesComponent.tsx
  
Implemented with specific messages

- CreateMeterModalComponent.tsx
  - NOTE: after failure message, the input fields are not reset. Need to make sure that resetState() is called after failure message.
- EditMeterModalComponent.tsx
- CreateUserModalComponent.tsx
  - NOTE: used a Promise to force the failure message to pop up
- EditUserModalComponent.tsx
  - NOTE: used a Promise to force the failure message to pop up
- ReadingsCSVUploadComponent.tsx
  - NOTE: not TODO DEBUG was added, didn’t know how to test SUCCESS NOTIFICATION (assuming it behaves the same as the failure notification which I could reproduce)
- MetersCSVUploadComponent.tsx
  - NOTE: not TODO DEBUG was added, didn’t know how to test SUCCESS NOTIFICATION (assuming it behaves the same as the failure notification which I could reproduce)
 
Not implemented
- CreateGroupModalComponent.tsx
- EditGroupModalComponent.tsx
- CreateConversionModalComponent.tsx
- EditConversionModalComponent.tsx
  - NOTE: after a fail save, the input fields are not reset. Need to make sure that resetState() (or similar function) is called after failure message.

With all of the edit/create files now investigated, I am going to start working on the 'Not implemented' files and reimplement the notification implementation that was used on the other edit/create files that have it implemented already.

**Comment by huss** [comment]
@aduques It has been a while since we communicated about this issue. Could you give me a status update? Thanks.

**Comment by aduques** [comment]
@huss Hello, I apologize about the delay and lack of communication. I picked up the issue again in December due to time conflicts with my semester. I have been having difficulty trying to figure out how to test the failure notifications on the PreferencesComponent.tsx (Admin Settings Page). I have tried creating a new function call for the 'submit' button that will try to pass an invalid entry.

	// TODO DEBUG: Force an error for testing
	const handleSubmitPreferences = (data: PreferenceRequestItem) => {
		// TODO DEBUG: Force an error for testing
		const testState = { ...data, displayTitle: '' };
		
		// Just return the mutation - don't unwrap or handle here
		return submitPreferences(testState);
		// return submitPreferences(data);  // Use this after testing
	};

...

				</Button>
				<Button
					type='submit'
					onClick={() => {
						// TODO DEBUG: Force an error for testing
						handleSubmitPreferences(localAdminPref)
						//submitPreferences(localAdminPref)
							.unwrap()
							.then(() => {
								showSuccessNotification(translate('updated.preferences'));
							})
							.catch(err => {
								showErrorNotification(
									translate('failed.to.submit.changes') + '"' + err.data + '"');
							})
					}}
					disabled={!hasChanges || Object.values(invalidFuncs).some(check => check())}
					color='primary'
				>

However, even with this, the attempt to overwrite the current change with an invalid one is not picked up. I was not able to identify what I could be overlooking. I believe it could be related to the UnsavedWarningComponent as it also has a separate function call to submit changes. Due to time constraints, I have not been able to investigate further.

As for the other files that need the failure/success notifications, I believe that I have been able to implement it correctly. To make further progress, I plan to prepare the pull request with the changes on the files excluding the PreferencesComponent.tsx, so that the current changes can be reviewed and can determine if the implementations should be done a different way.

**Comment by huss** [comment]
@aduques Thanks for the update. I'm happy to look at this once the PR is in. From trying to look at the code you gave, I see the displayTitle was set to an empty string. I think that is valid (see src/server/routes/preferences.js). If it was a number (without quotes so it isn't a string) then I think that might be invalid. Not sure if this will help.

**Comment by huss** [comment]
PR #1566 addressed this issue so it is being closed.
